src/components/TodoContext.tsx

```diff
@@ -0,0 +1,169 @@
+import React, {
+  useCallback,
+  useEffect,
+  useMemo,
+  useReducer,
+  useState,
+} from 'react';
+import { Todo } from '../types/Todo';
+import { deleteTodo, getTodos } from '../api/todos';
+
+export enum FilterBy {
+  All = 'ALL',
+  Active = 'ACTIVE',
+  Completed = 'COMPLETED',
+}
+
+export type FilterAction = FilterBy.All | FilterBy.Active | FilterBy.Completed;
+
+type TodoContextType = {
+  todos: Todo[];
+  originalTodos: Todo[];
+  dispatch: (action: Action) => void;
+  filteredBy: FilterBy;
+  handleFilterBy: (type: FilterBy) => void;
+  error: string | null;
+  handleError: (message: string) => void;
+  deleteCandidates: number[];
+  handleDeleteCandidates: (ids: number[]) => void;
+};
+
+export const TodoContext = React.createContext<TodoContextType>({
+  todos: [],
+  originalTodos: [],
+  dispatch: () => {},
+  filteredBy: FilterBy.All,
+  handleFilterBy: () => {},
+  error: null,
+  handleError: () => {},
+  deleteCandidates: [],
+  handleDeleteCandidates: () => {},
+});
+
+export type Props = {
+  children: React.ReactNode;
+};
+
+export enum ActionNames {
+  Load = 'LOAD',
+  Add = 'ADD',
+  Delete = 'DELETE',
+  Update = 'UPDATE',
+  ToggleCompleted = 'TOGGLE_COMPLETED',
+  ToggleAllCompleted = 'TOGGLE_ALL_COMPLETED',
+  ClearCompleted = 'CLEAR_COMPLETED',
+}
+
+type T = {
+  type: ActionNames.ToggleCompleted;
+  payload: { id: number; completed: boolean };
+};
+
+export type Action =
+  | { type: ActionNames.Load; payload: Todo[] }
+  | { type: ActionNames.Add; payload: Todo }
+  | { type: ActionNames.Delete; payload: number }
+  | { type: ActionNames.Update; payload: { id: number; title: string } }
+  | T
+  | { type: ActionNames.ToggleAllCompleted; payload: Todo[] }
+  | { type: ActionNames.ClearCompleted };
+
+export const Errors = {
```

Comment:
```suggestion
export const errors = {
```
it's just object, not a type (enum)